base/compiler/tfuncs.jl

```diff
@@ -902,10 +910,46 @@ function _getfield_tfunc_const(@nospecialize(sv), name::Const, setfield::Bool)
     return nothing
 end
 
-function _getfield_tfunc(@nospecialize(s00), @nospecialize(name), setfield::Bool)
-    if isa(s00, Conditional)
+function _getfield_tfunc(@specialize(lattice::InferenceLattice), @nospecialize(s00), @nospecialize(name), setfield::Bool)
```

Comment:
We probably want to define `_getfield_tfunc(@specialize(lattice::AbstractLattice), @nospecialize(s00), @nospecialize(name), setfield::Bool) = _getfield_tfunc(widenlattice(lattice), s00, name, setfield)` since otherwise we need to provide `_getfield_tfunc` implementation for every custom lattice?

Reply:
We could do that, but I kinda like the MethodError telling me I forgot to implement it. It's not in general true that you can just strip off the lattice element and pass it down.

Reply:
I'm fine with that for the meanwhile, but I guess it could be annoying if we need to provide many `tfunc` implementations in order to make the extended lattice to work fully.

I think we can setup additional interface for the lattice system like 
```julia
"""
    striplattice(lattice::L, @nospecialize(x)) where L<:AbstractLattice

Strips off additional lattice information from `x` that's specific to `L`.
"""
```

And provide default implementation of tfuncs like:
```julia
_getfield_tfunc(@specialize(lattice::AbstractLattice), @nospecialize(s00), @nospecialize(name), setfield::Bool) = _getfield_tfunc(widenlattice(lattice), striplattice(lattice, s00), striplattice(lattice, name), striplattice(lattice, setfield))
```
?

Reply:
I think that can just be spelled `widenlattice(L, var)` and yes, I think it's a useful operation - I'm just concerned that it's not generally correct.